app/code/community/FrontCommerce/Payline/Helper/Duplication.php

```diff
@@ -102,11 +106,187 @@ public function getDataToken($quote)
     }
   }
 
+
+  /**
+   * Source: Monext_Payline_IndexController::cptReturnAction()
+   * and also parts of FrontCommerce_Integration_Model_Api2_Sales_Cart_Order::_placeOrder
+   */
+  public function placeOrderFromPaylineToken(
+    $paylineToken,
+    $customer = null
+  ) {
+    $tokenData = Mage::getModel('payline/token')->load($paylineToken, 'token')->getData();
+
+    // Order is loaded from id associated to the token
+    if (sizeof($tokenData) == 0) {
+      Mage::helper('payline/logger')->log('[cptReturnAction] - token ' . $paylineToken . ' is unknown');
+      throw new \RuntimeException('Token ' . $paylineToken . ' is unknown');
+      return;
+    }
+    $this->order = Mage::getModel('sales/order')->loadByIncrementId($tokenData['order_id']);
+
+    if (!in_array($tokenData['status'], array(0, 3))) { // order update is already done => exit this function
+      $acceptedCodes = array(
+        '00000', // Credit card -> Transaction approved
+        '02500', // Wallet -> Operation successfull
+        '02501', // Wallet -> Operation Successfull with warning / Operation Successfull but wallet will expire
+        '04003', // Fraud detected - BUT Transaction approved (04002 is Fraud with payment refused)
+        '00100',
+        '03000',
+        '34230', // signature SDD
+        '34330' // prélèvement SDD
+      );
+
+      if (in_array($tokenData['result_code'], $acceptedCodes)) {
+        return [
+          'order_id' => $this->order->getIncrementId(),
+          'result_code' => $tokenData['result_code']
+        ];
+      } else {
+        throw new \RuntimeException(Mage::helper('payline')->__('Your payment is refused'));
+      }
+    }
+
+    $tokenForUpdate = Mage::getModel('payline/token')->load($tokenData['id']);
+    $webPaymentDetails = Mage::helper('payline')->initPayline('CPT')->getWebPaymentDetails(array('token' => $paylineToken, 'version' => Monext_Payline_Helper_Data::VERSION));
+
+    $this->order->getPayment()->setAdditionalInformation('payline_payment_info', $webPaymentDetails['payment']);
+
+    $paymentStatus = static::STATUS_ERROR;
+    $userMessage = null;
+    if (isset($webPaymentDetails)) {
+      if (is_array($webPaymentDetails) and !empty($webPaymentDetails['transaction'])) {
```

Comment:
there are way too many indent levels, I think those 2 conditions could easily be
inverted to return early and improve a bit the readability